### DISCUSSION TITLE
Predict mode for generating output without observed target variable

### DISCUSSION THREAD

**Discussion by shamshaw** [neuralhydrology/neuralhydrology#107] [Ideas]
Hi,

I'm wondering if you all have considered adding a predict mode or end point to generate predictions from a trained model without having the observed target variable in the input files? Right now, I've been adding essentially just a dummy observation column (just set all values to 1) to the input files where observed streamflow does not exist, which isn't too a big deal. However, it would be nicer to have a straightforward predict mode that just returns model predictions without invoking any evaluation. 

Thanks, Scott

**Comment by kratzert** [comment]
Hi Scott,

sorry, I must have missed this post. Yes, that is actually a good idea and I was myself ignored by this fact a couple of times. I think it shouldn't be too much work actually. E.g. the `_load_basin_data` method in the Dataset classes could implement a check that 

```python
if not self.is_train:
  for var in self.cfg.target_variables:
    if var not in df.columns:
      df[var] = np.nan
```

I think this would already do the trick. Maybe it would be best to do this in the `BaseDataset` class so that it doesn't have to be implemented for all dataset classes separately. I currently don't have much time but if you want, you could test it and open a PR?

**Reply by shamshaw** [reply]
Thanks for the reply and the suggested code -- that is essentially what I have implemented in a custom dataset class. I should be able to spend a little time seeing about putting it in the `BaseDataset` class and testing out. Will try to open a PR shortly

**Reply by shamshaw** [reply]
@kratzert Just following up on this and discussion in #121 
What I ended up doing is creating a `start_prediction` function in a `predict.py` placed under evaluation. Basically just lets me pass in a basin file and start and end date that I want to make predictions over and saves the results. Like one of you suggestions, I'm just temporarily updating the `test_basin_file`, `test_start_date`, and `test_end_date` in the config and then using the `Tester` class.

This I think will work well for our use case where we have a list of basin IDs that we'll want to make predictions in but will do that over changing date periods regularly (like when new data becomes available). I'm not sure if it makes sense to add something like this to neuralhydrology but just thought I would share what I ended up doing.

```
from pathlib import Path
import pickle

from neuralhydrology.evaluation import get_tester
from neuralhydrology.utils.config import Config

def start_prediction(cfg: Config, run_dir: Path, basin_file: str, start_date: str, end_date: str, output_dir: Path, epoch: int = None):
    """Start prediction for a list of basins over a defined date range using a trained network. 
    
    Useful for generation predictions at ungaged locations where no observations are available.  Does not assume observations are available so no metrics are saved

    Parameters
    ----------
    cfg : Config
        The run configuration, read from the run directory.
    run_dir : Path
        Path to the run directory.
    basin_file : str
        File name of basin list to make predictions
    start_date : str
        Start date in dd/mm/yyyy format for predictions
    end_date : str
        End date in dd/mm/yyyy format for predictions
    output_dir : Path
        Path to output directory for prediction results
    epoch : int, optional
        Define a specific epoch to evaluate. By default, the weights of the last epoch are used.

    """
    
    cfg.update_config(yml_path_or_dict = {'test_basin_file': basin_file,
                                          'test_start_date': start_date,
                                          'test_end_date': end_date})

    tester = get_tester(cfg=cfg, run_dir=run_dir, period = 'test', init_model=True)
    pred_results = tester.evaluate(epoch=epoch, save_results=False)

    file_name = output_dir / f"{Path(basin_file).stem}_prediction_results.p"
    file_name.parent.mkdir(parents=True, exist_ok=True)
    with file_name.open("wb") as fp:
        pickle.dump(pred_results, fp)
```

**Reply by kratzert** [reply]
Yes, that looks great. The few times I needed something similar this was basically what I implemented in an interactive environment. I am not sure though, if it is necessary to add this function but it is good to have this documented. In case that there are more requests like this in the future, we can either add it or at least point to this discussion.
